corehq/apps/reports/tests/test_readable_formdata.py

```diff
@@ -312,6 +315,67 @@ def test_get_readable_data_for_submission(self):
             json.dumps([FormQuestionResponse(q).to_json() for q in expected])
         )
 
+    @patch('corehq.apps.reports.formdetails.readable.get_questions')
+    def test_build_data_cleaning(self, questions_patch):
+        form_properties = OrderedDict()
+        form_properties['something'] = 'blue'
+        form_properties['lights'] = OrderedDict([('red', 'stop'), ('green', 'go')])
+        form_properties['cups_of_tea'] = [
+            {'details_of_cup': {'kind_of_cup': 'green'}},
+            {'details_of_cup': {'kind_of_cup': 'black'}},
+            {'details_of_cup': {'kind_of_cup': 'more green'}},
+        ]
+        form_properties['snacks'] = [
+            {'kind_of_snack': 'samosa'},
+            {'kind_of_snack': 'pakora'},
+        ]
+        formxml = FormSubmissionBuilder(
+            form_id='123',
+            form_properties=form_properties,
+        ).as_xml_string()
+
+        # Gross. Necessary because the test form isn't associated with an application.
```

Comment:
not sure if it would be any cleaner, but you could try using `XFormBuilder` for this as well. I had success with [a similar thing](https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/aggregate_ucrs/tests/test_aggregation.py#L129-L133) recently.